lib/elixir/lib/list.ex

```diff
@@ -467,62 +467,43 @@ defmodule List do
       iex> List.ascii_printable?('abc' ++ [0], 2)
       true
 
-  Improper lists are not printable, even if made only of ascii characters:
+  Last element in an improper list renders them not printable, even if made only of ASCII characters:
 
       iex> List.ascii_printable?('abc' ++ ?d)
       false
 
+      iex> List.ascii_printable?('abc' ++ ?d, 3)
+      true
+
   """
   @doc since: "1.6.0"
-  def ascii_printable?(list, counter \\ :infinity)
-
-  def ascii_printable?(_, 0) do
-    true
-  end
-
-  def ascii_printable?([char | rest], counter)
-      when is_integer(char) and char >= 32 and char <= 126 do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\n | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\r | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\t | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
-
-  def ascii_printable?([?\v | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
+  @spec ascii_printable?(maybe_improper_list, limit) :: boolean
+        when limit: :infinity | non_neg_integer
+  def ascii_printable?(list, limit \\ :infinity)
+      when is_list(list) and (limit == :infinity or (is_integer(limit) and limit >= 0)) do
+    ascii_printable_with_limit?(list, limit)
   end
 
-  def ascii_printable?([?\b | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
+  defp ascii_printable_with_limit?([], _limit),
+    do: true
 
-  def ascii_printable?([?\f | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
-  end
+  defp ascii_printable_with_limit?(_list, 0),
+    do: true
 
-  def ascii_printable?([?\e | rest], counter) do
-    ascii_printable?(rest, decrement(counter))
+  for control_char <- [?\a, ?\b, ?\e, ?\f, ?\n, ?\r, ?\t, ?\v] do
+    defp ascii_printable_with_limit?([unquote(control_char) | rest], limit),
+      do: ascii_printable_with_limit?(rest, decrement(limit))
   end
```

Comment:
It is not compiling (it was because I didn't run `make clean`).
How can I call a list comprehension in Erlang, from Elixir? or any why to bypass this limitation?

Reply:
This is a bootstrap issue. We can't use comprehensions on the List module because we likely need the List module for comprehensions to work (or something else that depends on the list). We can't be "smart" here.

But as I said in another comment, let's please keep the implementation as is in this commit and only change the minimum necessary to add guards (i.e. only rename the def/calls, no change to variables or syntax).

Reply:
done